### PR TITLE
Span parent cleanup

### PR DESCRIPTION
**Pull request by carlosalberto**
Addresses #261

### CODE REVIEW

File: api/src/main/java/io/opentelemetry/trace/Span.java [hunk 1]
```diff
@@ -304,6 +306,43 @@
    * @since 0.1.0
    */
   interface Builder {
+    /**
+     * Sets the parent {@code Span} to use. If not set, {@code Tracer.getCurrentSpan()} will be used
+     * as parent.
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/Span.java]: Does it worth commenting that the Tracer.getCurrentSpan() will be called during the build() time if no parent was specified? Otherwise some implementations may read it at the Builder creation time and reusing a SpanBuilder may generate undefined behaviors.

File: api/src/main/java/io/opentelemetry/trace/Tracer.java [hunk 2]
```diff
@@ -141,68 +141,17 @@
   Scope withSpan(Span span);
 
   /**
-   * Returns a {@link Span.Builder} to create and start a new child {@link Span} as a child of to
-   * the current {@code Span} if any, otherwise creates a root {@code Span}.
+   * Returns a {@link Span.Builder} to create and start a new child {@link Span}.
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/Tracer.java]: Not only child. As your comment from the Builder says you can also create root Spans. Maybe it makes sense to just say "a new Span"

File: opentracing-shim/src/main/java/io/opentelemetry/opentracingshim/SpanBuilderShim.java [hunk 3]
```diff
@@ -179,16 +179,14 @@ public SpanBuilder withStartTimestamp(long microseconds) {
 
   @Override
   public Span start() {
+    io.opentelemetry.trace.Span.Builder builder = tracer.spanBuilder(spanName);
 
-    io.opentelemetry.trace.Span.Builder builder;
     if (ignoreActiveSpan && parentSpan == null && parentSpanContext == null) {
-      builder = tracer.spanBuilderWithExplicitParent(spanName, null);
+      builder.setParent((io.opentelemetry.trace.Span) null);
```
Comment (bogdandrutu) [inline opentracing-shim/src/main/java/io/opentelemetry/opentracingshim/SpanBuilderShim.java]: You can now avoid caching the parent and redirect directly to the OpenTelemetry builder calls. Probably in a separate PR.

File: api/src/main/java/io/opentelemetry/trace/Span.java [hunk 4]
```diff
@@ -304,6 +306,43 @@
...
+     * process. This is only intended for use by RPC systems or similar.
+     *
+     * <p>If no {@link SpanContext} is available, users must call this method with a {@code null}
+     * remote parent {@link SpanContext}.
+     *
+     * <p>If called multiple times, only the last specified value will be used.
+     *
+     * @param remoteParent the {@link SpanContext} used as parent.
+     * @return this.
+     * @since 0.1.0
```
Comment (pavolloffay) [inline api/src/main/java/io/opentelemetry/trace/Span.java]: Add a reference to `Builder setParent(@Nullable Span parent);`. eg. `@see `

File: api/src/main/java/io/opentelemetry/trace/Span.java [hunk 5]
```diff
@@ -304,6 +306,43 @@
...
+     * <p>If called multiple times, only the last specified value will be used.
+     *
+     * @param parent the {@code Span} used as parent.
+     * @return this.
+     * @since 0.1.0
+     */
+    Builder setParent(@Nullable Span parent);
+
+    /**
+     * Sets the parent {@link SpanContext} to use. If not set, {@code Tracer.getCurrentSpan()} will
```
Comment (pavolloffay) [inline api/src/main/java/io/opentelemetry/trace/Span.java]: I would prefer keeping the main javadoc for  `Builder setParent(@Nullable Span parent);` , referencing it from here and just adding what is different (if there is something different).

File: api/src/main/java/io/opentelemetry/trace/Span.java (hunk 1 above)
Comment (carlosalberto) [reply api/src/main/java/io/opentelemetry/trace/Span.java]: Oh, yeah, good clarification.

File: opentracing-shim/src/main/java/io/opentelemetry/opentracingshim/SpanBuilderShim.java (hunk 3 above)
Comment (carlosalberto) [reply opentracing-shim/src/main/java/io/opentelemetry/opentracingshim/SpanBuilderShim.java]: Well, I've jumped from code entirely redirecting call to this, and not sure we will even keep this behavior now ;) (for example, if we have to support Out-of-band Spans, we might need to change the code again).

I will do such change later on, once we are sure things won't change as much ;)

File: api/src/main/java/io/opentelemetry/trace/Tracer.java (hunk 2 above)
Comment (carlosalberto) [reply api/src/main/java/io/opentelemetry/trace/Tracer.java]: "a new Span" works. Will update.

File: api/src/main/java/io/opentelemetry/trace/Span.java (hunk 1 above)
Comment (bogdandrutu) [reply api/src/main/java/io/opentelemetry/trace/Span.java]: Not sure I see any update about this.

File: api/src/main/java/io/opentelemetry/trace/Span.java [hunk 6]
```diff
@@ -304,6 +306,43 @@
...
+     * be used as parent.
+     *
+     * <p>This <b>must</b> be used to create a {@code Span} when the parent is in a different
+     * process. This is only intended for use by RPC systems or similar.
+     *
+     * <p>If no {@link SpanContext} is available, users must call this method with a {@code null}
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/Span.java]: Can you add a test for setParent(null). I think that will not compile.

Comment (yurishkuro) [reply api/src/main/java/io/opentelemetry/trace/Span.java]: I don't follow this - why MUST users call this with null?  If the intention is to force-start a new trace rather than use currentSpan, then I would prefer some other approach than setParent(null), maybe a default SpanContext constant.

Also, how is `setParent(null)` even going to compile given that the method is overloaded? You'd have to do something like `setParent((Span) null)`

Comment (carlosalberto) [reply api/src/main/java/io/opentelemetry/trace/Span.java]: That's correct, that won't compile - See the OT shim part, as I had to do `setParent((Span)null))`. This makes me thing we should have totally have them with different names, i.e. `setRemoteParent(SpanContext)`

> I don't follow this - why MUST users call this with null?

Actually that's something that nice to have, similar to the `ignoreActiveSpan()` we had in OT. We can call this `ignoreCurrentSpan()`?

Comment (pavolloffay) [reply api/src/main/java/io/opentelemetry/trace/Span.java]: I like more `.newTrace()` than `ignoreCurrentSpan()`. It seems more semantically correct if we want to provide an API to start a new trace.

Comment (carlosalberto) [reply api/src/main/java/io/opentelemetry/trace/Span.java]: So personally I'd like to see `setRootSpan()` or `setAsRootSpan()` to keep the style. Anyway, lets comment about that in the comments section ;)

**Comment by carlosalberto** [comment]
Hey hey - I've updated the code, adding `Span.Builder.ignoreCurrentSpan()` to define root Spans. The remaining changes to happen, in my opinion, are:

1. Do we rename the overload taking a `SpanContext` to `setRemoteParent()`? I'd strongly say YES.
2. `ignoreCurrentSpan()` is a good name? @pavolloffay suggested `newTrace()`, and from my side I like `setAsRootSpan()`. 
2. How do we handle the `null` cases for `setParent()`? My personal choice would be to completely ignore that case, and have users use explicitly call `ignoreCurrentSpan()`.

Based on the result of this, I will update this PR later today ;)

File: api/src/main/java/io/opentelemetry/trace/Span.java (hunk 6 above)
Comment (yurishkuro) [reply api/src/main/java/io/opentelemetry/trace/Span.java]: I like `newTrace()` too

File: api/src/main/java/io/opentelemetry/trace/NoopTrace.java [hunk 7]
```diff
@@ -121,6 +105,21 @@ public Span startSpan() {
       return BlankSpan.INSTANCE;
     }
 
+    @Override
+    public NoopSpanBuilder setParent(@Nullable Span parent) {
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/NoopTrace.java]: this should not accept null anymore.

Comment (carlosalberto) [reply api/src/main/java/io/opentelemetry/trace/NoopTrace.java]: Agreed. And to be clear: now users *should* use the `ignoreCurrentSpan()` method (or `newTrace()` or whatever name we give to it). Makes sense?

File: api/src/main/java/io/opentelemetry/trace/Span.java (hunk 6 above)
Comment (carlosalberto) [reply api/src/main/java/io/opentelemetry/trace/Span.java]: Thanks for the feedback. So, you might not like it, but trying to keep the current convention, we ought to use `setNewTrace()`. Works for you?

File: api/src/main/java/io/opentelemetry/trace/NoopTrace.java (hunk 7 above)
Comment (bogdandrutu) [reply api/src/main/java/io/opentelemetry/trace/NoopTrace.java]: Yes.

File: api/src/main/java/io/opentelemetry/trace/Span.java (hunk 6 above)
Comment (bogdandrutu) [reply api/src/main/java/io/opentelemetry/trace/Span.java]: setNoParent()? This will also make sense for Tags/Baggages.

**Comment by carlosalberto** [comment]
Hey hey - I've updated the code to reflect the latest feedback.

1. `setNoParent()` for now. I personally also like this a little bit more, as it aligns with `setParent()`. Let me know if you think this is not a good-enough choice.
2. `setParent()` receiving `null` is considered an error BUT it can still receive `SpanContext.INVALID`, and it will behave as before (defining a root `Span`). I think this will keep us consistent with what we had before. Else, let me know.

File: api/src/main/java/io/opentelemetry/trace/Span.java [hunk 8]
```diff
@@ -304,6 +305,53 @@
...
+
+    /**
+     * Sets the parent {@link SpanContext} to use. If not set, the value of {@code
+     * Tracer.getCurrentSpan()} at {@link #startSpan()} time will be used as parent.
+     *
+     * <p>Similar to {@link #setParent(Span parent)} but this <b>must</b> be used to create a {@code
+     * Span} when the parent is in a different process. This is only intended for use by RPC systems
+     * or similar.
+     *
+     * <p>If no {@link SpanContext} is available, users must call {@link #setNoParent()} in order to
```
Comment (pavolloffay) [inline api/src/main/java/io/opentelemetry/trace/Span.java]: You should document that `null` is invalid - based on your noop implementation.

Comment (carlosalberto) [reply api/src/main/java/io/opentelemetry/trace/Span.java]: Well, that's why I wanted to keep a few things 'repeated' in this and the other method's docs - it might seem too verbose to have many things repeated, but given that semantically they are not the same, this could be handy.

So I I will also restore that the last specified value will be used.

**Comment by carlosalberto** [comment]
> I don't like If null in setParent(SpanContext) is invalid and if extract can return a null context.

IIRC the idea was that `extract()` can return a no-op `SpanContext` (we don't expose this anymore, but the implementation can have it internally). In this case, this won't be an error.

Unless we **really** want to allow user to pass `null` only for this overload.

**Comment by bogdandrutu** [comment]
Did we not decide that extract never returns null?

**Comment by carlosalberto** [comment]
@bogdandrutu Correct. Maybe we should make this super clear on the docs of `extract()`? (in a different issue/PR).

**Comment by carlosalberto** [comment]
Updated the documentation to reflect's @pavolloffay feedback.

For the sake of keeping speed (and @bogdandrutu being traveling at the moment), I will likely merge later today or early tomorrow. Feel free to open an issue/PR in case of some required/desired change ;)

File: api/src/main/java/io/opentelemetry/trace/Span.java [hunk 9]
```diff
@@ -304,6 +305,56 @@
...
+     *
+     * <p>It is an error to specify a {@code null} parent.
+     *
+     * @param remoteParent the {@link SpanContext} used as parent.
+     * @return this.
+     * @see #setParent(Span parent)
+     * @see #setNoParent()
+     * @since 0.1.0
+     */
+    Builder setParent(SpanContext remoteParent);
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/Span.java]: Just add a `@throw NullPointerException if ...` to make sure users see this. Maybe we can remove the `<p>It is an error to specify a {@code null} parent.`

**Comment by pavolloffay** [comment]
I think the extract should be better documented and explicitly ban `null` as returned value.

**Comment by carlosalberto** [comment]
Merging. @pavolloffay I will create an Issue for tracking the improvement for extraction.
